Move mobile specific CPUCachingAllocator to c10/mobile folder.

kimishpatel opened the pull request:
Summary: Plus add some more comments about the usage, limitations and cons.

Test Plan: Build and run benchmark binary.

Differential Revision: D23944193

gchanan commented on c10/mobile/CPUCachingAllocator.h:
```diff
@@ -10,6 +10,39 @@
+/*
+ * CPUCachingAllocator:
+ * Why?
+ *    It has been observed that on mobile some platforms, such as pixel 3, returns
```
nit: did you mean: "It has been observed that some mobile platforms, such as pixel 3", return ..."

gchanan commented on c10/mobile/CPUCachingAllocator.h:
```diff
@@ -10,6 +10,39 @@
 #include <c10/util/flat_hash_map.h>
 
+/*
+ * CPUCachingAllocator:
```
nit: can you put a brief summary of the disclaimer first, like:

`DISCLAIMER: this is subject to change (beta) and only supported on mobile builds.  See below for more information"

kimishpatel committed:
Move mobile specific CPUCachingAllocator to c10/mobile folder. (#45364)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45364

Plus add some more comments about the usage, limitations and cons.

Test Plan: Build and run benchmark binary.

Reviewed By: gchanan

Differential Revision: D23944193

fbshipit-source-id: 25d91f4ca0ede4c48cb96faff3a830f1260b41b2
